target-chains/near/receiver/src/governance.rs

```diff
@@ -34,162 +29,221 @@ use {
-        let action = cursor.read_u8()?.try_into()?;
-        let target = cursor.read_u16::<BigEndian>()?;
-
-        assert!(module == GOVERNANCE_MODULE);
-        assert!(target == 0 || target == u16::from(WormholeChain::Near));
-        assert!(magic == u32::from_le_bytes(GOVERNANCE_MAGIC));
-
-        Ok(match action {
-            ActionId::ContractUpgrade => {
-                let mut hash = [0u8; 32];
-                cursor.read_exact(&mut hash)?;
-                Self::ContractUpgrade(hash)
-            }
+        // Verify Target is this contract.
+        let (input, target) = be_u8(input)?;
+        if target != GOVERNANCE_MODULE {
+            return Err(Error::InvalidGovernanceModule);
+        }
 
-            ActionId::SetDataSources => {
-                let mut sources = Vec::new();
-                let count = cursor.read_u8()?;
-
-                for _ in 0..count {
-                    let mut emitter = [0u8; 32];
-                    cursor.read_exact(&mut emitter)?;
-                    sources.push(Source {
-                        emitter,
-                        pyth_emitter_chain: Chain::from(WormholeChain::from(
-                            cursor.read_u16::<BigEndian>()?,
-                        )),
-                    });
-                }
+        // Safely parse the action ID. [ref:action_discriminants]
+        let (input, action) = be_u8(input)?;
+        let action = GovernanceActionId::from_u8(action).ok_or(Error::InvalidGovernanceAction)?;
+
+        match action {
+            GovernanceActionId::UpgradeContract => {
+                let (_input, bytes) = all_consuming(take(32usize))(input)?;
+                let mut codehash = [0u8; 32];
+                codehash.copy_from_slice(bytes);
 
-                Self::SetDataSources(sources)
+                // Make sure the buffer has been fully consumed.
+                Ok(GovernanceAction::UpgradeContract { codehash })
             }
 
-            ActionId::SetGovernanceSource => {
-                let mut emitter = [0u8; 32];
-                cursor.read_exact(&mut emitter)?;
-                Self::SetGovernanceSource(Source {
-                    emitter,
-                    pyth_emitter_chain: Chain(cursor.read_u16::<BigEndian>()?),
+            GovernanceActionId::AuthorizeGovernanceDataSourceTransfer => {
+                let (_input, claim_vaa) = all_consuming(take(input.len()))(input)?;
+                Ok(GovernanceAction::AuthorizeGovernanceDataSourceTransfer {
+                    claim_vaa: claim_vaa.to_vec(),
                 })
             }
 
-            ActionId::SetStalePriceThreshold => {
-                let stale_price_threshold = cursor.read_u64::<BigEndian>()?;
-                Self::SetStalePriceThreshold(stale_price_threshold)
+            GovernanceActionId::SetDataSources => {
+                let (_input, data_sources) = many0(|input| {
+                    let (_, (chain, emitter)) = all_consuming(|input| {
```

Comment:
the nesting of the many_0 and all_consuming here confuses me. shouldn't it be the other way around?

Reply:
Yes though is actually incorrect either way around, fixed.